### PR TITLE
Add Loader component

### PR DESCRIPTION
**Pull request by remybach**
🐿 v2.12.3

## Feature Description

Loader component for use in all of our forms. Includes JS util for controlling it.

Note: I've had to do a fair bit of hacking around in the demos/test code to make it so that I could use `@partial-block` because I felt that defining named inline partials wasn't as neat. At least now we have the ability to do so going forward 😄

## Link to Ticket / Card:

https://trello.com/c/0xuGAx9L/1074-5-loaders-on-all-the-pages

## Screenshots:

Note: the example where data is passed in is incorrect here (`title=title`). I've fixed it since making the gif.

### CODE REVIEW

File: demos/data.json [hunk 1]
```diff
@@ -3,6 +3,10 @@
     "filterList": ["GBR", "JPN", "USA", "FRA"],
     "value": "USA"
   },
+  "loader": {
+    "title": "Hooray!",
+    "@partial-block": "Your content is busy loading! <sup>(not really, this is just a demo)</sup>"
```
Comment (liamkeaton) [inline demos/data.json]: What is partial block?

**Comment by remybach** [comment]
To answer your other question first, it's standard handlebars: https://handlebarsjs.com/partials.html#partial-block

I'd really rather not use those inline partials... it makes the implementation uglier and more complicated for no reason. They're only really necessary if you want to define multiple different bits of marked up content.

I'd much rather have this:

```markdown
{{#> n-conversion-forms/partials/loader }}
  <p>Loading the thing.</p>
{{/ n-conversion-forms/partials/loader }}
```

... than this:

```markdown
{{#> n-conversion-forms/partials/loader }}
  {{#*inline "content"}}
	<p>Loading the thing.</p>
  {{/inline}}
{{/ n-conversion-forms/partials/loader }}
```

I will admit that there's some sunk cost thinking going on here as I've already gone through the pain of making it all work nicely 😅, but I don't see the harm in enabling the use of this. Especially since it's standard and neater/easier for the end user.

**Comment by liamkeaton** [comment]
I really don't like having to replicate parts of the templating process like this -

https://github.com/Financial-Times/n-conversion-forms/pull/126/files#diff-d5458fb0964573729ea7e6d7ee0672fdR41

I feel it's going to trip us up in future, if it's not needed i.e. using the inline partials, then why take the risk 🤷‍♂️

**Comment by remybach** [comment]
To turn that around on you 😬: If using inline partials is not needed (which in this case it isn't _needed_), then why make it a worse experience for the user?

The example you provided is entirely harmless: if that `#if` isn't found, the data isn't mutated. The only way I see that tripping us up in the future is if we upgraded handlebars entirely _and_ if they decided to completely change how partial blocks work. That would be a huge undertaking of its own that would naturally involve refactoring that file.

**Comment by liamkeaton** [comment]
You are right, inline blocks make it slightly messier for the end user so I don't mind enabling inline block. My main worries are -

* two ways in `ncf` to do inlining of content (why does handlebars have two ways!)
* string replace on our templates before render (I may have a fix)

File: styles/loader.scss [hunk 2]
```diff
@@ -0,0 +1,44 @@
+/* stylelint-disable selector-no-qualifying-type */
```
Comment (remybach) [inline styles/loader.scss]: Disabling this because it disallows classes that represent state (such as `.is-visible` below).

File: demos/main.js [hunk 3]
```diff
@@ -0,0 +1,10 @@
+const Loader = require('../utils/loader');
+
+if (document.querySelector('.ncf__loader')) {
+	const loader = new Loader(document);
+	loader.show();
+}
+
+// Make the textarea display all the rows it's given.
+const $textarea = document.getElementById('example-code');
+$textarea.rows = ($textarea.value.match(/\n/gi).length) + 1;
```
Comment (liamkeaton) [inline demos/main.js]: Oh nice 👌

File: styles/loader.scss (hunk 2 above)
Comment (liamkeaton) [reply styles/loader.scss]: Why has this been enabled?

Comment (remybach) [reply styles/loader.scss]: No idea, but it must have come from downstream.

It's more to try prevent uses like `a.foo` or `div#bar` from what I've gathered.

Comment (liamkeaton) [reply styles/loader.scss]: Shame you can't turn off for a particular line

**Pull request merged**